rest/route/distro.go

```diff
@@ -17,33 +17,229 @@ import (
 	"github.com/pkg/errors"
 )
 
+///////////////////////////////////////////////////////////////////////
+//
+// GET /rest/v2/distros/{distro_id}/setup
+
+type distroIDGetSetupHandler struct {
+	distroID string
+	sc       data.Connector
+}
+
+func makeGetDistroSetup(sc data.Connector) gimlet.RouteHandler {
+	return &distroIDGetSetupHandler{
+		sc: sc,
+	}
+}
+
+func (h *distroIDGetSetupHandler) Factory() gimlet.RouteHandler {
+	return &distroIDGetSetupHandler{
+		sc: h.sc,
+	}
+}
+
+// Parse fetches the distroId from the http request.
+func (h *distroIDGetSetupHandler) Parse(ctx context.Context, r *http.Request) error {
+	h.distroID = gimlet.GetVars(r)["distro_id"]
+
+	return nil
+}
+
+// Run returns the given distro's setup script.
+func (h *distroIDGetSetupHandler) Run(ctx context.Context) gimlet.Responder {
+	distro, err := h.sc.FindDistroById(h.distroID)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "Database error for find() by distro id '%s'", h.distroID))
+	}
+
+	script := &model.APIDistroSetup{}
+	if err = script.BuildFromService(*distro); err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "API error converting from distro.Distro.Setup to model.APIDistroSetup"))
+	}
+
+	return gimlet.NewJSONResponse(script)
+}
+
+///////////////////////////////////////////////////////////////////////
+//
+// PATCH /rest/v2/distros/{distro_id}/setup
+
+type distroIDChangeSetupHandler struct {
+	Setup    string
+	distroID string
+	sc       data.Connector
+}
+
+func makeChangeDistroSetup(sc data.Connector) gimlet.RouteHandler {
+	return &distroIDChangeSetupHandler{
+		sc: sc,
+	}
+}
+
+func (h *distroIDChangeSetupHandler) Factory() gimlet.RouteHandler {
+	return &distroIDChangeSetupHandler{
+		sc: h.sc,
+	}
+}
+
+// Parse fetches the distroId and JSON payload from the http request.
+func (h *distroIDChangeSetupHandler) Parse(ctx context.Context, r *http.Request) error {
+	h.distroID = gimlet.GetVars(r)["distro_id"]
+	body := util.NewRequestReader(r)
+	defer body.Close()
+
+	if err := util.ReadJSONInto(body, h); err != nil {
```

Comment:
should we have the read the body into the distro struct which we have as a field in the handler type?

Reply:
As discussed